src/Microsoft.SymbolStore/ChecksumValidator.cs

```diff
@@ -0,0 +1,125 @@
+﻿using Microsoft.FileFormats.PE;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Security.Cryptography;
+using System.Text;
+
+namespace Microsoft.SymbolStore
+{
+    class ChecksumValidator
+    {
+        private const string pdbStreamName = "#Pdb";
+        private const uint pdbIdSize = 20;
+
+        internal static void Validate(ITracer tracer, Stream pdbStream, IEnumerable<PdbChecksum> pdbChecksums)
+        {
+            using (MemoryStream ms = new MemoryStream())
+            {
+                pdbStream.CopyTo(ms);
+                byte[] bytes = ms.ToArray();
```

Comment:
I believe this method of copying the stream will result in two different in-memory copyies of all the bytes. I think you want:

```
        byte[] bytes = new byte[pdbStream.Length];
        if (pdbStream.Read(bytes, offset: 0, count: bytes.Length) != bytes.Length)
        {
            throw ...
        }
```

Reply:
looks better, thanks

Reply:
Just been testing this and I think my code was depending on there being two different copies (one with zeroed bytes for the checksum and the unaltered stream for writing to disk)

I'm getting `System.BadImageFormatException: Invalid COR20 header signature.` when I try and load the pdb and I think this is probably the reason why.

Reply:
as an aside - why do we need to zero the signature bytes for the checksum? Are they not deterministic?

Reply:
Reverted this change and it works fine.

Reply:
Fixed by making a backup of the pdb Id and restoring before writing to disk to save making a copy of the whole stream.